library/core/src/mem/maybe_uninit.rs

```diff
@@ -1284,3 +1279,44 @@ impl<T> MaybeUninit<T> {
         }
     }
 }
+
+impl<T, const N: usize> MaybeUninit<[T; N]> {
+    /// Transposes a `MaybeUninit<[T; N]>` into a `[MaybeUninit<T>; N]`.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(maybe_uninit_uninit_array_transpose)]
+    /// # use std::mem::MaybeUninit;
+    ///
+    /// let data: [MaybeUninit<u8>; 1000] = MaybeUninit::uninit().transpose();
+    /// ```
+    #[unstable(feature = "maybe_uninit_uninit_array_transpose", issue = "96097")]
+    pub fn transpose(self) -> [MaybeUninit<T>; N] {
+        // SAFETY: T and MaybeUninit<T> have the same layout
+        let result = unsafe { super::transmute_copy(&self) };
+        super::forget(self);
+        result
```

Comment:
suggestion: while this works fine, since none of it can panic, we try in the docs to follow the "pre-pit your peaches" order of things, rather than "forget it afterwards", and consider `forget` soft-deprecated.

So please update this to something like 
```rust
    pub fn transpose(self) -> [MaybeUninit<T>; N] {
        // SAFETY: T, ManuallyDrop<T>, and MaybeUninit<T> have the same layout
        unsafe { super::transmute_copy(&ManuallyDrop::new(self)) }
    }
```

(And ditto for the one below)

Reply:
Oh yeah, that's so much better. Is there a reason `forget` isn't deprecated? It seems like you could always replace it with `let _ = ManuallyDrop::new(foo);` (or `drop(ManuallyDrop::new(foo));`) worse case scenario.

Reply:
Yeah, you can -- in fact that's literally what `forget` does

https://github.com/rust-lang/rust/blob/b8c35ca26b191bb9a9ac669a4b3f4d3d52d97fb1/library/core/src/mem/mod.rs#L153-L155

I think it's not deprecated because, **un**like `mem::uninitialized`, there's lots of code that used `forget` that's still perfectly correct because it only does things (like `copy_nonoverlapping`) that can't panic.

But maybe it's time to start thinking about it, now that `ManuallyDrop` has been around for long enough.  Or at least a deprecated-in-rust-2.0 for now so it gets picked up by https://github.com/rust-lang/rust/blob/9b0a099dfc9a97ecd10adb319396c731c4b2d169/compiler/rustc_lint_defs/src/builtin.rs#L2243-L2258

Reply:
Hmm, maybe not.  It looks like `mem::forget(guard);` is still very common in the library for panic guards, where there's no need to consume them like in the example, and thus the `forget` phrasing is kinda nice.

But maybe that also just means that the `PanicGuard` types should have a `defuse` method or something as part of its API instead of telling people to `forget` it.

Reply:
Hmmm, yeah I see the appeal for panic guards, but poking around the stdlib I think it's actually pretty hard to understand what's going on unless you know what the guards are for, so maybe encouraging people to write a `defuse` method or add a comment explaining the weirdness is a good thing. Created #103101.